modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMapper.java

```diff
@@ -0,0 +1,563 @@
+                            // just bubble up failed mapping, we will deal with it on top of the tree
+                            return res;
+                        }
+                    } catch (ColocationMappingException e) {
+                        return new FailedMapping(new FragmentMappingException(e.getMessage(), input, e));
+                    }
+                }
+
+                return res;
+            }
+        }
+
+        private Mapping mapTrimExchange(IgniteTrimExchange rel) {
+            RelNode input = rel.getInput();
+
+            Mapping mapping = computeMapping(input);
+
+            try {
+                return mapComputableSource(rel.sourceId()).colocate(mapping);
+            } catch (ColocationMappingException e) {
+                return new FailedMapping(new FragmentMappingException(e.getMessage(), input, e));
+            }
+        }
+
+        private Mapping mapTableModify(IgniteTableModify rel) {
+            RelNode input = rel.getInput();
+            Mapping mapping = computeMapping(input);
+
+            if (mapping.failed()) {
+                // just bubble up failed mapping, we will deal with it on top of the tree
+                return mapping;
+            }
+
+            IgniteTable igniteTable = rel.getTable().unwrapOrThrow(IgniteTable.class);
+
+            ExecutionTarget target = targets.get(igniteTable.id());
+            assert target != null : "No colocation group for " + igniteTable.id();
+
+            try {
+                return newMapping(-1, target).colocate(mapping);
+            } catch (ColocationMappingException e) {
+                return new FailedMapping(new FragmentMappingException(e.getMessage(), input, e));
+            }
+        }
+
+        private Mapping mapComputableSource(long sourceId) {
+            ExecutionTarget target = context.targetFactory().someOf(context.nodes());
+
+            return newMapping(sourceId, target);
+        }
+
+        private Mapping mapTableScan(long sourceId, ProjectableFilterableTableScan rel) {
+            IgniteTable igniteTable = rel.getTable().unwrapOrThrow(IgniteTable.class);
+
+            ExecutionTarget target = targets.get(igniteTable.id());
+            assert target != null : "No colocation group for " + igniteTable.id();
+
+            return newMapping(sourceId, target);
+        }
+    }
+
+    private Mapping newMapping(long sourceId, ExecutionTarget target) {
+        return new ColocatedMapping(LongSets.singleton(sourceId), target);
+    }
+
+    private static Mapping combineMappings(List<Mapping> mappings) {
+        return new CombinedMapping(mappings);
+    }
+
+    private interface Mapping {
+        boolean failed();
+
+        boolean colocated();
+
+        Mapping colocate(Mapping other) throws ColocationMappingException;
+
+        List<ColocationGroup> createColocationGroups() throws FragmentMappingException;
+    }
+
+    private static class FailedMapping implements Mapping {
```

Comment:
1. this this is refactoring only, I rely on existing tests. But I've filed an [issue](https://issues.apache.org/jira/browse/IGNITE-20504) to revise and improve test coverage
2. technically, yes. There is nothing wrong to have both FailedMapping and ColocatedMapping within the same CombinedMapping. But to avoid unnecessary work, mapping is checked if it's failed